docs: add documentation and migration instructions for new composability API (#4072)
backstage/backstage: docs/plugins/composability.md

intermediate components in the app, for example like this:

```tsx
const appRoutes = (
  <Routes>
    <Route path="/foo" element={<FooPage />} />
    <Route path="/bar" element={<BarPage />} />
  </Routes>
);

const App = () => (
  <AppProvider>
    <AppRouter>
      <Root>{appRoutes}</Root>
    </AppRouter>
  </AppProvider>
);
```

### New Routing System

A big piece of what is enabled by moving over to this new composability system
is to make `RouteRef`s useful. The `RouteRef`s no longer have their own path, in
fact the only required parameter is currently a `title`. Instead of assigning a
path to each `RouteRef` and possibly overriding these paths in the app, the
concrete `path` for each `RouteRef` is discovered based on the element tree in
the app. Let's consider the following example:

```tsx
const appRoutes = (
  <Routes>
    <Route path="/foo" element={<FooPage />} />
    <Route path="/bar" element={<BarPage />} />
  </Routes>
);
```

We'll assume that `FooPage` and `BarPage` are routable extensions, exported by
the `fooPlugin` and `barPlugin` respectively. Since the `FooPage` is a routable
extension it has a `RouteRef` assigned as its mount point, which we'll refer to

> as `fooRootRouteRef`.

Reviewer: probably clearer for the reader to call it `fooPageRouteRef` etc

Given the above example, the `fooRootRouteRef` will be associated with the
`'/foo'` route. The path is no longer accessible via the `path` property of the
`RouteRef` though, as the routing structure is tied to the app's react tree. We
instead use the new `useRouteRef` hook if we want to create a concrete link to
the page. The `useRouteRef` hook takes a single `RouteRef` as its only
parameter, and returns a function that is called to create the URL. For example
like this:

```tsx
const MyComponent = () => {
  const fooRoute = useRouteRef(fooRouteRef);
  return <a href={fooRoute()}>Link to Foo</a>;
};
```

Now let's assume that we want to link from the `BarPage` to the `FooPage`.
Before the introduction of the new composability system, we would do this by

> importing the `fooRootRouteRef` from the `fooPlugin`. This created an

Reviewer: ```suggestion
importing the `fooRootRouteRef` exported by the `fooPlugin`. This created an
```

makes it a bit clearer, i read "import from" as "add an import statement in" :)

and instead have to be associated with a target route using route bindings in
the app.

We create a new `ExternalRouteRef` inside the `barPlugin`, using a neutral name
that describes its role in the plugin rather than a specific plugin page that it
might be linking to, allowing the app to decide the final target. If the
`BarPage` for example wants to link to an external page in the header, it might
declare an `ExternalRouteRef` similar to this:

```ts
const headerLinkRouteRef = createExternalRouteRef();
```

### Binding External Routes in the App

The association of external routes is controlled by the app. Each
`ExternalRouteRef` of a plugin should be bound to an actual `RouteRef`, usually
from another plugin. The binding process happens once at app startup, and is
then used through the lifetime of the app to help resolve concrete route paths.

Using the above example of the `BarPage` linking to the `FooPage`, we might do
something like this in the app:

```ts
createApp({
  bindRoutes({ bind }) {
    bind(barPlugin.externalRoutes, {
      headerLink: fooPlugin.routes.root,
    });
  },
});
```

Given the above binding, using `useRouteRef(headerLinkRouteRef)` within the

> `barPlugin` will let us create a link whatever path the `FooPage` is mounted at.

Reviewer: ```suggestion
`barPlugin` will let us create a link to whatever path the `FooPage` is mounted at.
```

Note that we are not importing and using the `RouteRef`s directly in the app,
and instead rely on the plugin instance to access routes of the plugins. This is
a new convention that was introduced to provide better namespacing and
discoverability of routes, as well as reduce the number of separate exports from
each plugin package. The route references would be supplied to `createPlugin`
like this:

```ts
// In foo-plugin
export const fooPlugin = createPlugin({
  routes: {
    root: fooRootRouteRef,
  },
  ...
})

// In bar-plugin
export const barPlugin = createPlugin({
  externalRoutes: {
    headerLink: headerLinkRouteRef,
  },
  ...
})
```

Also note that you almost always want to create the route references themselves
in a different file than the one that creates the plugin instance, for example a
top-level `routes.ts`. This is to avoid circular imports when you use the route

> references from other parts of the app.

Reviewer: ```suggestion
references from other parts of the same plugin.
```

right? or am i confused? 🤔

```tsx
// Creation of a parameterized route
const myRouteRef = createRouteRef({
  title: 'My Named Route',
  params: ['name']
})

// In the app, where MyPage is a routable extension with myRouteRef set as mountPoint
<Route path='/my-page/:name' element={<MyPage />}/>

// Usage within a component
const myRoute = useRouteRef(myRouteRef)
return (
  <div>
    <a href={myRoute({name: 'a'})}>A</a>
    <a href={myRoute({name: 'b'})}>B</a>
  </div>
)
```

It is currently not possible to have parameterized `ExternalRouteRef`s, or to
bind an external route to a parameterized route, although this may be added in
the future if needed.

### New Catalog Components

The established pattern for selecting what plugins should be available on each
catalog page is to use custom components in the app, with logic embedded in the
render function. Typically this takes form as a component that either receives
the entity via props or uses the `useEntity` hook to retrieve the selected
entity. A `switch` or `if` / `else if` chain is then used to select what
children should be rendered based on information in the entity.

This pattern will no longer work with the new composability system, and in

> general is very difficult to build any form declarative model around, as it

Reviewer: ```suggestion
general is very difficult to build any form of declarative model around, as it
```

depends on runtime execution. To help replace existing code, a new
`EntitySwitch` component has been added to the `@backstage/catalog` plugin,

> which grabs the selected entity from context, and selects at most one element to

Reviewer: ```suggestion
which grabs the selected entity from a context, and selects at most one element to
```

> render using a list of `EntitySwitch.Case`s children.

Reviewer: ```suggestion
render using a list of `EntitySwitch.Case` children.
```

`FlatRoutes` from `@backstage/core`. As opposed to the `Routes` component,
`FlatRoutes` only considers the first level of `Route` components in its
children, and provides any additional children to the outlet of the route. It
also removes the need to append `"/*"` to paths, as it is added automatically.

```diff
const AppRoutes = () => (
-  <Routes>
+  <FlatRoutes>
    ...
-    <Route path="/docs/*" element={<DocsRouter />} />
+    <Route path="/docs" element={<DocsRouter />} />
    ...
-  </Routes>
+  </FlatRoutes>
);
```

The next step should be to switch from using `EntityPageLayout` to
`EntityLayout`, as this can also be done without waiting for plugins to be
ported. You should also replace the top-level `Router` from the catalog plugin
with the separate `CatalogIndexPage` and `CatalogEntityPage` extensions that
have been added to the catalog:

```diff
-<Route
-  path={`${catalogRouteRef.path}/*`}
-  element={<CatalogRouter EntityPage={EntityPage} />}
-/>
+<Route path="/catalog" element={<CatalogIndexPage />} />
+<Route
+  path="/catalog/:namespace/:kind/:name"
+  element={<CatalogEntityPage />}
+>
+  <EntityPage />

> +</Route>

Reviewer: This one has both element and children - mistake? if not, need to explain what that is about

Author: It's explained above. Was a bit brief about it though as I didn't want this to dive too deep into how react router works.

> `FlatRoutes` only considers the first level of `Route` components in its
children, and provides any additional children to the outlet of the route.

which has a slightly different pattern for expressing the contents and paths.

Porting from the old to the new API is just a matter of moving some things
around. For example, given the following existing code:

```tsx
<EntityPageLayout>
  <EntityPageLayout.Content
    path="/"
    title="Overview"
    element={<ComponentOverviewContent entity={entity} />}
  />
  <EntityPageLayout.Content
    path="/sentry"
    title="Sentry"
    element={<SentryRouter entity={entity} />}
  />
  <EntityPageLayout.Content
    path="/kubernetes/*"
    title="Kubernetes"
    element={<KubernetesRouter entity={entity} />}
  />
</EntityPageLayout>
```

It would be ported to this:

```tsx
<EntityLayout>
  <EntityLayout.Route path="/" title="Overview">
    <ComponentOverviewContent entity={entity} />
  </EntityLayout.Route>

  <EntityLayout.Route path="/sentry" title="Sentry">
    <SentryRouter entity={entity} />
  </EntityLayout.Route>

  <EntityLayout.Route path="/kubernetes" title="Kubernetes">
    <KubernetesRouter entity={entity} />
  </EntityLayout.Route>
</EntityLayout>
```

In addition to the renaming, the `element` prop has been moved to `children`.

> Also note that the `/*` suffix has been remove from the `"/kubernetes"` path, as

Reviewer: ```suggestion
Also note that the `/*` suffix has been removed from the `"/kubernetes"` path, as
```

---
id: composability
title: Composability System Migration

> description:
>   Documentation and migration instructions for new composability APIs.

Reviewer 2: ```suggestion
# prettier-ignore
description: Documentation and migration instructions for new composability APIs.
```

For SEO, Docusaurus needs it in one line. But prettier wants to split. https://github.com/backstage/backstage/pull/3927

in place to support the new composability system.

### Component Data

Component data is a new composability primitive that is introduced as a way to
provide a new data dimension for React components. Data is attached to React
components using a key, and is then readable from any JSX elements created with
those components, using the same key, as illustrated by the following example:

```tsx
const MyComponent = () => <h1>This is my component</h1>;
attachComponentData(MyComponent, 'my.data', 5);

const element = <MyComponent />;
const myData = getComponentData(element, 'my.data');
// myData === 5
```

The purpose of component data is to provide a method for embedding data that can
be inspected before rendering elements. Element inspection is a pattern that is
quite common among React libraries, and used for example by `react-router` and
`material-ui` to discover properties of the child elements before rendering.
Although in those libraries only the element type and props are typically
inspected, while our component data adds more structured access and simplifies
evolution by allowing for multiple different versions of a piece of data to be
used and interpreted at once.

> The initial use-cases for component data is support route and plugin discovery
> through elements in the app. Through this we allow for the React element tree in

Reviewer 2: ```suggestion
The initial use-case for component data is to support route and plugin discovery
through elements in the app. Through this we allow for the React element tree in
```

Before the introduction of the new composability system, we would do this by
importing the `fooRootRouteRef` from the `fooPlugin`. This created an
unnecessary dependency on the plugin, and also provided little flexibility in
allowing the app to tie plugins together, with the links instead being dictated
by the plugins themselves. To solve this, we introduce `ExternalRouteRef`s. Much
like regular route references, they can be passed to `useRouteRef` to create
concrete URLs, but they can not be used as mount points in routable component
and instead have to be associated with a target route using route bindings in
the app.

We create a new `ExternalRouteRef` inside the `barPlugin`, using a neutral name
that describes its role in the plugin rather than a specific plugin page that it
might be linking to, allowing the app to decide the final target. If the
`BarPage` for example wants to link to an external page in the header, it might
declare an `ExternalRouteRef` similar to this:

```ts
const headerLinkRouteRef = createExternalRouteRef();
```

### Binding External Routes in the App

The association of external routes is controlled by the app. Each

> `ExternalRouteRef` of a plugin should be<- bound to an actual `RouteRef`,

Reviewer 2: ```suggestion
`ExternalRouteRef` of a plugin should be bound to an actual `RouteRef`,
```

> `ExternalRouteRef` of a plugin should be<- bound to an actual `RouteRef`,

Reviewer 3: `<-` some leftover charachters?

> usually from another plugin. The binding process happens once att app startup,

Author: ```suggestion
usually from another plugin. The binding process happens once at app startup,
```

general is very difficult to build any form declarative model around, as it
depends on runtime execution. To help replace existing code, a new
`EntitySwitch` component has been added to the `@backstage/catalog` plugin,
which grabs the selected entity from context, and selects at most one element to
render using a list of `EntitySwitch.Case`s children.

For example, if you want all entities of kind `"Template"` to be rendered with a
`MyTemplate` component, and all other entities to be rendered with a `MyOther`
component, you would do the following:

```tsx
<EntitySwitch>
  <EntitySwitch.Case if={isKind('template')}>
    <MyTemplate />
  </EntitySwitch.Case>

  <EntitySwitch.Case>
    <MyTemplate />
  </EntitySwitch.Case>
</EntitySwitch>

// Shorter form if desired:
<EntitySwitch>
  <EntitySwitch.Case if={isKind('template')} children={<MyTemplate />}/>
  <EntitySwitch.Case children={<MyTemplate />}/>
</EntitySwitch>
```

The `EntitySwitch` component will render the children of the first
`EntitySwitch.Case` that returns `true` when the selected entity is passed to
the function of the `if` prop. If none of the cases match, no children will be

> rendered, and if a case doesn't specify an `if` filter function, it will always

Reviewer 4: Does this effectively act as a default / fallback case?

Author: Yep! But it's not required, in case you don't want to render anything if there are no matches

up from context inside the plugin instead. See the sections below for how to
carry out migrations of some common entity page patterns.

Once the app element tree doesn't contain any intermediate components, and all
plugin imports have been switched to extensions rather than plain components,
the app has been fully ported.

### Switching from EntityPageLayout to EntityLayout

The existing `EntityPageLayout` is replaced by the new `EntityLayout` component,
which has a slightly different pattern for expressing the contents and paths.

Porting from the old to the new API is just a matter of moving some things
around. For example, given the following existing code:

```tsx
<EntityPageLayout>
  <EntityPageLayout.Content
    path="/"
    title="Overview"
    element={<ComponentOverviewContent entity={entity} />}
  />
  <EntityPageLayout.Content
    path="/sentry"
    title="Sentry"
    element={<SentryRouter entity={entity} />}
  />
  <EntityPageLayout.Content
    path="/kubernetes/*"
    title="Kubernetes"
    element={<KubernetesRouter entity={entity} />}
  />
</EntityPageLayout>
```

It would be ported to this:

```tsx
<EntityLayout>
  <EntityLayout.Route path="/" title="Overview">

>     <ComponentOverviewContent entity={entity} />}

Reviewer 4: Extra `}` at end

>     <ComponentOverviewContent entity={entity} />}

Author: ```suggestion
    <ComponentOverviewContent entity={entity} />
```

  </EntityLayout.Route>

  <EntityLayout.Route path="/sentry" title="Sentry">

>     <SentryRouter entity={entity} />}
>   </EntityLayout.Route>
> 
>   <EntityLayout.Route path="/kubernetes" title="Kubernetes">
>     <KubernetesRouter entity={entity} />}

Author: ```suggestion
    <SentryRouter entity={entity} />
  </EntityLayout.Route>

  <EntityLayout.Route path="/kubernetes" title="Kubernetes">
    <KubernetesRouter entity={entity} />
```
